labrad/decorators.py

```diff
@@ -21,202 +21,167 @@
-
-        elif Nparams == 1:
-            accepts_s = params.get(args[0], [])
-            accepts_t = [T.parseTypeTag(s) for s in accepts_s]
-
-            if Nrequired == 0:
-                # if accepted types were specified, add '' to the list
-                # we don't add '' if the list of accepted types is empty,
-                # since this would make '' the ONLY accepted type
-                if len(accepts_t) and T.LRNone() not in accepts_t:
-                    accepts_s.append(': defaults [%s=%r]' \
-                                     % (args[0], defaults[0]))
-                    accepts_t.append(T.LRNone())
-
-                @wraps(f)
-                def handleRequest(self, c, data):
-                    if data is None:
-                        return f(self, c)
-                    return f(self, c, data)
-
+        if Nrequired > 1:
+            self.expand_cluster = "always"
+        elif Nparams > 1:
+            self.expand_cluster = "optional"
+        else: # one or fewer arguments
+            self.expand_cluster = "never"
+
+        self.allow_none = Nrequired == 0
+
+        if Nparams:
+            for tag in params[args[0]]:
+                tt = T.parseTypeTag(tag)
+                if isinstance(tt, T.LRNone) and Nrequired != 1:
+                    raise ValueError("First argument {} cannot accept '' "
+                          "unless other arguments are optional".format(args[0]))
+                if isinstance(tt, (T.LRAny, T.LRCluster)) and self.expand_cluster=="optional":
+                    raise ValueError("First argument {} cannot accept type {} "
+                          "because other arguments are optional".format(args[0], tt))
+            for arg in args[1:]:
+                for tag in params[arg]:
+                    if isinstance(T.parseTypeTag(tag), T.LRNone):
+                        raise ValueError("Argument {} cannot accept ''".format(arg))
+
+        # Build list of accepted data types.
+        # This is basically every combination of accepted types for each agrument,
+        # including omitting any number of trailing optional arguments.
+
+        accepted_types = []
+        for i in range(Nrequired, Nparams+1):
+            if i == 0:
+                accepted_types.append('_')
             else:
-                # nothing special to do here
-                handleRequest = f
-
+                accept_tuples = itertools.product(*[params[arg] for arg in args[:i]])
+                accepted_types.extend(''.join(x) for x in accept_tuples)
+        self.accepts = accepted_types
+
+    def __get__(self, obj, objtype):
+        """Allow direct call on settings from within the Server code
+
+        Uses the python descriptor protocol.
+        """
+        # When called on an instance, we return the setting function
+        # to allow direct calling.  When called on the class we return
+        # the Setting object so that the server code can call
+        # handleRequest
```

Comment:
We should add a test that this works, e.g. a new setting in test_server.py that we call from one of the tests.